lib/analyzer/image-inspector.ts

```diff
@@ -152,17 +155,116 @@ async function pullImage(
       platform,
     );
     if (pullAndSaveSuccessful) {
-      return;
+      imageNames.push(targetImage);
+      return imageNames;
     }
   }
 
-  await pullFromContainerRegistry(
+  const { indexDigest, manifestDigest } = await pullFromContainerRegistry(
     docker,
     targetImage,
     imageSavePath,
     username,
     password,
   );
+
+  const imageNameWithIndexDigest = getImageNameWithIndexDigest(
+    indexDigest,
+    targetImage,
+  );
+  const imageNameWithManifestDigest = getImageNameWithManifestDigest(
+    manifestDigest,
+    targetImage,
+  );
+
+  return getImageNames(
+    imageNameWithIndexDigest,
+    imageNameWithManifestDigest,
+    targetImage,
+  );
+}
+
+export function getImageNames(
+  indexImageName: string | undefined,
+  manifestImageName: string | undefined,
+  targetImage: string,
+): string[] {
+  const imageNames: Set<string> = new Set();
+
+  if (indexImageName) {
+    imageNames.add(indexImageName);
+  }
+  if (manifestImageName) {
+    imageNames.add(manifestImageName);
+  }
+  if (indexImageName !== targetImage && manifestImageName !== targetImage) {
+    imageNames.add(targetImage);
+  }
+  return Array.from(imageNames);
+}
+
+export function getImageNameWithIndexDigest(
+  indexDigest: string | undefined,
+  targetImage: string,
+): string | undefined {
+  let imageNameWithIndexDigest: string | undefined;
+  const { imageName, imageTag, imageDigest } = getImageParts(targetImage);
+
+  if (!indexDigest) {
+    imageNameWithIndexDigest = undefined;
+  }
+
+  // nginx:tag or nginx@sha256:manifestdigest or nginx@sha256:indexdigest
+  if (indexDigest && (imageTag || imageDigest)) {
+    imageNameWithIndexDigest = `${imageName}@${indexDigest}`;
+  }
+
+  return imageNameWithIndexDigest;
+}
+
+export function getImageNameWithManifestDigest(
+  manifestDigest: string | undefined,
+  targetImage: string,
+): string | undefined {
+  let imageNameWithManifestDigest: string | undefined;
+  const { imageName, imageTag, imageDigest } = getImageParts(targetImage);
+
+  if (!manifestDigest) {
+    imageNameWithManifestDigest = undefined;
+  }
+
+  // nginx:tag or nginx@sha256:indexdigest or nginx@sha256:manifestdigest
+  if (manifestDigest && (imageTag || imageDigest)) {
+    imageNameWithManifestDigest = `${imageName}@${manifestDigest}`;
+  }
+
+  return imageNameWithManifestDigest;
+}
+
+export function getImageParts(imageWithTag: string): {
+  imageName: string;
+  imageTag: string;
+  imageDigest: string;
+} {
+  // we're matching pattern: <registry:port_number>(optional)/<image_name>(mandatory):<image_tag>(optional)@<tag_identifier>(optional)
+  // extracted from https://github.com/docker/distribution/blob/master/reference/regexp.go
```

Comment:
As written in Slack, I had the following comment in a local branch for exactly this regex:
```
  // this regex has been copied from
  // https://github.com/distribution/distribution/blob/fb2188868d771aa27e5781a32bf78d4c113c18a6/reference/regexp.go#L101
  // (code has been modified to print the regex), and then adjusted to
  // Javascript. The required modifications were replacing `[:xdigit:]` with
  // `[a-fA-F0-9]` and escaping the slashes.
  // Note that the digest matched in this Regex will match digests that have
  // uppercase-letters, while the regex used in validateDigest does *not* match
  // uppercase-letters. This simply matches the behaviour from the upstream
  // `reference` and `go-digest `packages.
  //
  // This Regex contains three capture groups:
  // 1) The repository / image name
  // 2) tag
  // 3) digest
```

But also, my regex was *slightly* different 😅 
```
/^((?:(?:(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?|\[(?:[a-fA-F0-9:]+)\])(?::[0-9]+)?\/)?[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:\/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?)(?::([\w][\w.-]{0,127}))?(?:@([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][a-fA-F0-9]{32,}))?$/g;
```

(the comment mentions `validateDigest`; a function that made sure that the digest was also correct. There is this slight difference between the digest validation and the regex, for reasons I couldn't figure out)

Reply:
I am not a regex expert either, but I have tested the possible match here

Your regex https://regex101.com/r/cpokY6/1, it might be started even with the image name. It can't match the full `nginx`. I tried the debugger, and it was showing red starting there, I don't know why though 😅

The regex in Kubernetes Monitor https://regex101.com/r/E91cld/1, it matches all the case 😄

Reply:
Interesting, because I thought I tested that and it seemed to work 🤔 Well, good that yours is working then! :)